orttraining/orttraining/training_api/include/optimizer.h

```diff
@@ -56,8 +56,8 @@ struct Optimizer {
   // Initialize an optimizer module from an ORT inference session with loaded
   // training ONNX model For each parameter, initialize the OptimizerState based
   // on the graph input's ValueInfoProto if the parameter doesn't have it already.
-  Optimizer(const std::string& optim_path_or_bytes,
-            const std::unordered_map<std::string, std::shared_ptr<Parameter>>& parameters);
+  Optimizer(const std::unordered_map<std::string, std::shared_ptr<Parameter>>& named_parameters,
```

Comment:
Now that I am reading this, I am wondering why the optimizer takes in a const ref to named_parameters when these parameters are going to be updated? I guess it is fine considering that the parameter itself is a shared pointer.

But this really means that the named_parameters unordered map is not shared between user, module and optimizer. It will be copy constructed for the optimizer (and it seems will also be copied for the module). So any changes made to the unordered map (like adding a new entry in the map), will not reflect in the user's checkpoint state. Only changes to the shared pointer parameter will be reflected in the users state. Is that what we want?

Reply:
I dont think that is what we want... Do we expect to make any changes to the unordered map itself? Will the number of entries or the keys ever change?

Reply:
This is an interesting question. As synced offline, the case is very rare (and to be honest, I never see such a case yet, e.g. optimizer or other components update/insert new values in the map). So at least currently it is fine we don't need do it. 

Another perspective - optimizer or other components can update weights values, but don't own the parameter lists.

Reply:
I am guessing this is perhaps not required, but if there is a specific order required to instantiate the optimizer, we should document it in the constructor in the code.
We should also take in a const ref to the unordered_map in the module constructor (rn it is not a const ref) implying to the user that changes made to the map are not for consumption to the user/optimizer.